web/packages/teleport/src/Discover/Server/EnrollEc2Instance/EnrollEc2Instance.tsx

```diff
@@ -0,0 +1,284 @@
+        setFetchEc2InstancesAttempt({ status: 'success' });
+        setTableData({ ...data, fetchStatus: 'disabled' });
+        return;
+      }
+
+      // Check if fetched EC2 instances are already in the cluster
+      // so that they can be disabled in the table.
+
+      // Builds the predicate string that will query for
+      // all the fetched EC2 instances by searching by the AWS instance ID label.
+      const instanceIdPredicateQueries: string[] = fetchedEc2Instances.map(
+        d =>
+          `labels["teleport.dev/instance-id"] == "${d.awsMetadata.instanceId}"`
+      );
+      const fullPredicateQuery = instanceIdPredicateQueries.join(' || ');
+      const { agents: fetchedNodes } = await nodeService.fetchNodes(
+        cfg.proxyCluster,
+        {
+          query: fullPredicateQuery,
+          limit: fetchedEc2Instances.length,
+        }
+      );
+
+      const ec2InstancesLookupByInstanceId: Record<string, Node> = {};
+      fetchedNodes.forEach(
+        d => (ec2InstancesLookupByInstanceId[d.awsMetadata.instanceId] = d)
+      );
+
+      // Check for already existing EC2 instances.
+      const checkedEc2Instances: CheckedEc2Instance[] = fetchedEc2Instances.map(
+        ec2 => {
+          const instance =
+            ec2InstancesLookupByInstanceId[ec2.awsMetadata.instanceId];
+          if (instance) {
+            return {
+              ...ec2,
+              ec2InstanceExists: true,
+            };
+          }
+          return ec2;
+        }
+      );
+
+      setFetchEc2InstancesAttempt({ status: 'success' });
+      setTableData({
+        currRegion,
+        nextToken,
+        fetchStatus: nextToken ? '' : 'disabled',
+        items: [...data.items, ...checkedEc2Instances],
+      });
+    } catch (err) {
+      const errMsg = getErrMessage(err);
+      setTableData(data);
+      setFetchEc2InstancesAttempt({ status: 'failed', statusText: errMsg });
+      emitErrorEvent(`ec2 instance fetch error: ${errMsg}`);
+    }
+  }
+
+  async function fetchEc2InstanceConnectEndpoints() {
+    const integrationName = (agentMeta as NodeMeta).integration.name;
+
+    setFetchEc2IceAttempt({ status: 'processing' });
+    try {
+      const { endpoints: fetchedEc2Ices } =
+        await integrationService.fetchAwsEc2InstanceConnectEndpoints(
+          integrationName,
+          {
+            region: selectedInstance.awsMetadata.region,
+            vpcId: selectedInstance.awsMetadata.vpcId,
+          }
+        );
+      setFetchEc2IceAttempt({ status: 'success' });
+      return fetchedEc2Ices;
+    } catch (err) {
+      const errMsg = getErrMessage(err);
+      setFetchEc2InstancesAttempt({ status: 'failed', statusText: errMsg });
+      emitErrorEvent(`ec2 instance connect endpoint fetch error: ${errMsg}`);
+    }
+  }
+
+  function clear() {
+    setFetchEc2InstancesAttempt({ status: '' });
+    setTableData(emptyTableData);
+    setSelectedInstance(null);
+  }
+
+  function handleOnProceed() {
```

Comment:
In my head the flow should be:
- select ec2 instance
- create ec2 instance
- handle eice subflow (check if exists in `create-complete`, ...)

That's what we do with other resources, I think
Eg for the databases, we first create the Database and only then check if we have a database-server.

Reply:
The issue with this would be that then it's possible that we'd create a non-functional resource in Teleport. If we create the ec2 instance first and then they run into issues setting up the EICE and abort the flow, they would have a node in their cluster that doesn't work. It seems to me that it's a better UX to create the node after, is there any benefit to doing it before that I might not be aware of?

Reply:
That's ok :+1: 
Thank you for adding an explanation.

The flow I suggested is applied in databases, I think: we create the DB and then check/create connectivity